src/monodroid/jni/monodroid-glue.c

```diff
@@ -704,55 +738,191 @@ typedef void* (*mono_mkbundle_init_ptr) (void (*)(const MonoBundledAssembly **),
-	const char *v;
-	char *libapp;
+	mono_bool _dummy_needs_free = FALSE;
+	if (needs_free == NULL)
+		needs_free = &_dummy_needs_free;
 
-#ifndef RELEASE
-	libapp = path_combine (override_dirs [0], "libmonodroid_bundle_app.so");
+	*needs_free = FALSE;
+	char *dso_path = get_full_dso_path (base_dir, dso_name, needs_free);
+	if (file_exists (dso_path))
+		return dso_path;
 
-	if (file_exists (libapp))
-		return libapp;
+	*needs_free = FALSE;
+	free (dso_path);
+	return NULL;
+}
 
-	free (libapp);
-#endif
 
-	if (dir) {
-		v = (*env)->GetStringUTFChars (env, dir, NULL);
-		if (v) {
-			libapp = path_combine (v, "libmonodroid_bundle_app.so");
-			(*env)->ReleaseStringUTFChars (env, dir, v);
-			if (file_exists (libapp))
-				return libapp;
-		}
+static void
+dso_alloc_cleanup (char **dso_path, mono_bool *needs_free)
+{
+	if (dso_path != NULL) {
+		if (needs_free && *needs_free)
+			free (*dso_path);
+		*dso_path = NULL;
+	}
+
+	if (needs_free)
+		*needs_free = FALSE;
+}
+
+static char*
+get_full_dso_path_on_disk (const char *dso_name, mono_bool *needs_free)
+{
+	mono_bool _dummy_needs_free = FALSE;
+	if (needs_free == NULL)
+		needs_free = &_dummy_needs_free;
+
+	*needs_free = FALSE;
+	if (embedded_dso_mode)
+		return NULL;
+
+	char *dso_path = NULL;
+	for (int i = 0; i < MAX_OVERRIDES; i++) {
+		if (override_dirs [i] == NULL)
+			continue;
+		dso_path = get_existing_dso_path_on_disk (override_dirs [i], dso_name, needs_free);
+		if (dso_path != NULL)
+			return dso_path;
+		dso_alloc_cleanup (&dso_path, needs_free);
 	}
+
+	for (int i = 0; i < app_lib_directories_size; i++) {
+		dso_path = get_existing_dso_path_on_disk (app_lib_directories [i], dso_name, needs_free);
+		if (dso_path != NULL)
+			return dso_path;
+		dso_alloc_cleanup (&dso_path, needs_free);
+	}
+
 	return NULL;
 }
 
+// This function could be improved if we somehow marked an apk containing just the bundled app as
+// such - perhaps another __XA* environment variable? Would certainly make code faster.
+static void
+setup_bundled_app (const char *dso_name)
+{
+	static int dlopen_flags = RTLD_LAZY;
+	void *libapp = NULL;
+
+	if (embedded_dso_mode) {
+		log_info (LOG_DEFAULT, "bundle app: embedded DSO mode");
+		libapp = load_dso_from_any_directories (dso_name, dlopen_flags);
+	} else {
+		mono_bool needs_free = FALSE;
+		log_info (LOG_DEFAULT, "bundle app: normal mode");
+		char *bundle_path = get_full_dso_path_on_disk (dso_name, &needs_free);
+		log_info (LOG_DEFAULT, "bundle_path == %s", bundle_path ? bundle_path : "<NULL>");
+		if (bundle_path == NULL)
+			return;
+		log_info (LOG_BUNDLE, "Attempting to load bundled app from %s", bundle_path);
+		libapp = load_dso (bundle_path, dlopen_flags, TRUE);
+	}
+
+	if (libapp == NULL) {
+		log_info (LOG_DEFAULT, "No libapp!");
+		if (!embedded_dso_mode) {
+			log_fatal (LOG_BUNDLE, "bundled app initialization error");
+			exit (FATAL_EXIT_CANNOT_LOAD_BUNDLE);
+		} else {
+			log_info (LOG_BUNDLE, "bundled app not found in the APK, ignoring.");
```

Comment:
Why should we ignore here and not exit?  Why would `embedded_dso_mode` matter for this check?

Reply:
Short version: if we exit, embedded DSO stops working.

Long version: in non-embedded mode the decision to set up a bundled app is based on the presence, on the filesystem, of the bundled app `.so` file - we `stat` the filesystem to make the decision. In embedded mode we cannot do the stat inside the zip/apk and so we always `dlopen` all the `.so` and see if that succeeded. Since it would always fail here, if the application wasn't a bundled one, the DSO mode would never work. As it currently stands, there's no other way to make this work except to ignore the failure to load the bundle DSO. It can (and should) be improved (as stated in the comment above the `setup_bundled_app` function) by allowing the code to detect if we're inside a bundled app by other means - for instance an environment variable added to the environment file when building the app. This change, however, is beyond scope of this PR.